In [679]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
import matplotlib.pylab as plt
%matplotlib inline
from sklearn import preprocessing
from sklearn import model_selection
from keras.models import Sequential
from keras.layers import Dense, Embedding, LSTM, GRU
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.utils import to_categorical
from scikeras.wrappers import KerasClassifier, KerasRegressor
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.utils import resample
import pickle

# Multiclass classification models

In this notebook I will create classification models to label the years as per their economic activity. According to bibliography, the world has experience 4 recessions (decrease of GDP) and 4 downturns (low GDP growth) in the last 70 years. I will label them accordingly:

1. Recession-peak: 1975, 1982, 1991, 2009, 2020
2. Recovery: 2 years after each crisis
3. Global-downturn: 1958, 1998, 2001, 2012
4. Expansion: all other years

## IMF dataset

This is the dataset containing global data for 1980-2027.

In [683]:
orignal_imf = pd.read_csv(r'C:\Users\luana\Ironhack DA\Unit 9\final_bootcamp_project\csv_files\imf_clean.csv')
orignal_imf

WEO Subject Code             Country  \
0              NGDP_R             Albania   
1           NGDP_RPCH             Albania   
2                NGDP             Albania   
3               NGDPD             Albania   
4              PPPGDP             Albania   
...               ...                 ...   
4040          D_NGDPD  Sub-Saharan Africa   
4041               DS  Sub-Saharan Africa   
4042         DS_NGDPD  Sub-Saharan Africa   
4043              DSI  Sub-Saharan Africa   
4044        DSI_NGDPD  Sub-Saharan Africa   

                               Subject Descriptor  \
0         Gross domestic product, constant prices   
1         Gross domestic product, constant prices   
2          Gross domestic product, current prices   
3          Gross domestic product, current prices   
4          Gross domestic product, current prices   
...                                           ...   
4040                         External debt, total   
4041            External debt, total debt service   
4042            External debt, total debt service   
4043  External debt, total debt service, interest   
4044  External debt, total debt service, interest   

                                          Subject Notes  \
0     Expressed in billions of national currency uni...   
1     Annual percentages of constant price GDP are y...   
2     Expressed in billions of national currency uni...   
3     Values are based upon GDP in national currency...   
4     These data form the basis for the country weig...   
...                                                 ...   
4040                                               None   
4041                                               None   
4042                                               None   
4043                                               None   
4044                                               None   

                                               Units     Scale  \
0                                  National currency  Billions   
1                                     Percent change      None   
2                                  National currency  Billions   
3                                       U.S. dollars  Billions   
4     Purchasing power parity; international dollars  Billions   
...                                              ...       ...   
4040                                  Percent of GDP      None   
4041                                    U.S. dollars  Billions   
4042                                  Percent of GDP      None   
4043                                    U.S. dollars  Billions   
4044                                  Percent of GDP      None   

                          Country/Series-specific Notes     1980     1981  \
0     Source: IMF Staff Estimates. Official national...  311.514  329.270   
1     See notes for:  Gross domestic product, consta...    2.684    5.700   
2     Source: IMF Staff Estimates. Official national...   18.489   19.126   
3     See notes for:  Gross domestic product, curren...    1.946    2.229   
4     See notes for:  Gross domestic product, curren...    5.759    6.663   
...                                                 ...      ...      ...   
4040                                               None   18.911   21.069   
4041                                               None   10.493   10.300   
4042                                               None    3.350    3.426   
4043                                               None    3.301    3.541   
4044                                               None    1.040    1.055   

         1982  ...      2019      2020      2021      2022      2023  \
0     338.819  ...   837.786   808.617   877.475   912.574   935.388   
1       2.900  ...     2.088    -3.482     8.516     4.000     2.500   
2      19.698  ...  1691.900  1644.080  1889.840  2059.020  2176.660   
3       2.296  ...    15.399    15.161    18.310    18.256    18.842   
4       7.280  ...    41.623    40.658  

In [681]:
orignal_imf['Scale'].unique()

array(['Billions', 'None', 'Units', 'Millions'], dtype=object)

### Preparing the dataset

In [684]:
# I need to rearrange the dataset to make it suitable for the classification
# I can drop all the categorical columns because they only give information about the indicators
# I will use data until 2020 for the model and then try to label the years 2021-2027 

imf = orignal_imf.select_dtypes(np.number)
imf

1980     1981     1982     1983     1984     1985     1986     1987  \
0     311.514  329.270  338.819  342.546  349.397  344.156  363.428  360.521   
1       2.684    5.700    2.900    1.100    2.000   -1.500    5.600   -0.800   
2      18.489   19.126   19.698   19.900   19.645   20.065   20.692   20.531   
3       1.946    2.229    2.296    2.319    2.290    2.339    2.587    2.566   
4       5.759    6.663    7.280    7.649    8.083    8.214    8.848    8.994   
...       ...      ...      ...      ...      ...      ...      ...      ...   
4040   18.911   21.069   21.791   29.217   38.232   41.154   41.957   38.565   
4041   10.493   10.300   11.741   13.934   15.031   16.869   17.096   17.030   
4042    3.350    3.426    3.512    5.457    7.972    8.582    7.109    6.116   
4043    3.301    3.541    3.144    3.457    3.234    3.887    4.041    4.385   
4044    1.040    1.055    0.929    1.109    1.307    1.721    1.569    1.567   

         1988     1989  ...      2018      2019      2020      2021      2022  \
0     355.474  390.310  ...   820.653   837.786   808.617   877.475   912.574   
1      -1.400    9.800  ...     4.019     2.088    -3.482     8.516     4.000   
2      20.238   22.228  ...  1636.730  1691.900  1644.080  1889.840  2059.020   
3       2.530    2.779  ...    15.157    15.399    15.161    18.310    18.256   
4       9.181   10.476  ...    40.055    41.623    40.658    45.953    51.189   
...       ...      ...  ...       ...       ...       ...       ...       ...   
4040   39.362   40.063  ...    39.422    41.042    43.811    40.804    39.774   
4041   17.986   16.911  ...   114.102   113.329   109.146   127.688   149.875   
4042    6.621    5.700  ...     7.230     6.637     6.380     7.232     7.655   
4043    4.650    4.894  ...    17.880    18.733    18.035    18.286    18.511   
4044    1.542    1.593  ...     1.021     1.063     1.083     0.975     0.909   

          2023         2024         2025         2026         2027  
0      935.388   965.321000   996.211000  1030.080000  1065.110000  
1        2.500     3.200000     3.200000     3.400000     3.400000  
2     2176.660  2287.470000  2403.260000  2530.790000  2664.390000  
3       18.842    19.858000    21.219000    22.528000    23.889000  
4       54.338    57.254000    60.184000    63.415000    66.840000  
...        ...          ...          ...          ...          ...  
4040    37.853    37.758550    37.758550    37.758550    37.758550  
4041   129.985    42.604425    42.604425    42.604425    42.604425  
4042     6.163     6.297775     6.297775     6.297775     6.297775  
4043    20.489     7.451600     7.451600     7.451600     7.451600  
4044     0.938     1.191625     1.191625     1.191625     1.191625  

[4045 rows x 48 columns]

In [685]:
# I need each year to be a row
imf = imf.transpose().reset_index()
imf

index         0      1         2       3       4        5          6  \
0   1980   311.514  2.684    18.489   1.946   5.759    5.935  116584.54   
1   1981   329.270  5.700    19.126   2.229   6.663    5.809  120786.15   
2   1982   338.819  2.900    19.698   2.296   7.280    5.814  121689.95   
3   1983   342.546  1.100    19.900   2.319   7.649    5.809  120446.72   
4   1984   349.397  2.000    19.645   2.290   8.083    5.623  120297.85   
..   ...       ...    ...       ...     ...     ...      ...        ...   
43  2023   935.388  2.500  2176.660  18.842  54.338  232.701  327239.88   
44  2024   965.321  3.200  2287.470  19.858  57.254  236.964  338749.82   
45  2025   996.211  3.200  2403.260  21.219  60.184  241.240  350720.75   
46  2026  1030.080  3.400  2530.790  22.528  63.415  245.688  363876.33   
47  2027  1065.110  3.400  2664.390  23.889  66.840  250.153  377613.26   

           7          8  ...      4035     4036      4037    4038      4039  \
0    5557.56    6919.41  ...   0.39000  0.70600  -6.29800  4.9720   60.0180   
1    5757.84    7016.15  ...  -0.52400  0.19700  -9.89200 -8.5170   66.1260   
2    5800.93    7074.57  ...  -1.24300 -0.55800 -11.62700 -2.0610   76.5560   
3    5741.66    6997.24  ...  -0.88400  0.62700  -9.15100 -0.2070   80.0050   
4    5734.57    6763.89  ...  -0.54900 -1.06200  -1.78900 -0.0870   83.2400   
..       ...        ...  ...       ...      ...       ...     ...       ...   
43  15599.44  761490.02  ... -40.27200  4.48500  -9.66100  1.0360  803.0300   
44  16148.12  802716.63  ... -13.42365 -4.79135   1.80165  4.4091  246.9388   
45  16718.77  846078.15  ... -13.42365 -4.79135   1.80165  4.4091  246.9388   
46  17345.89  894000.51  ... -13.42365 -4.79135   1.80165  4.4091  246.9388   
47  18000.73  944610.78  ... -13.42365 -4.79135   1.80165  4.4091  246.9388   

        4040        4041      4042     4043      4044  
0   18.91100   10.493000  3.350000   3.3010  1.040000  
1   21.06900   10.300000  3.426000   3.5410  1.055000  
2   21.79100   11.741000  3.512000   3.1440  0.929000  
3   29.21700   13.934000  5.457000   3.4570  1.109000  
4   38.23200   15.031000  7.972000   3.2340  1.307000  
..       ...         ...       ...      ...       ...  
43  37.85300  129.985000  6.163000  20.4890  0.938000  
44  37.75855   42.604425  6.297775   7.4516  1.191625  
45  37.75855   42.604425  6.297775   7.4516  1.191625  
46  37.75855   42.604425  6.297775   7.4516  1.191625  
47  37.75855   42.604425  6.297775   7.4516  1.191625  

[48 rows x 4046 columns]

In [686]:
#Column names from the indicators
columns = ['year']

for i in range(len(orignal_imf)):
    column_name = (orignal_imf['Country'][i]+'_'+orignal_imf['Subject Descriptor'][i])
    column_name = str(column_name).lower().replace(' ','_')
    columns.append(column_name)

len(columns)

4046

In [687]:
imf.columns= columns
imf.head()

year  albania_gross_domestic_product,_constant_prices  \
0  1980                                          311.514   
1  1981                                          329.270   
2  1982                                          338.819   
3  1983                                          342.546   
4  1984                                          349.397   

   albania_gross_domestic_product,_constant_prices  \
0                                            2.684   
1                                            5.700   
2                                            2.900   
3                                            1.100   
4                                            2.000   

   albania_gross_domestic_product,_current_prices  \
0                                          18.489   
1                                          19.126   
2                                          19.698   
3                                          19.900   
4                                          19.645   

   albania_gross_domestic_product,_current_prices  \
0                                           1.946   
1                                           2.229   
2                                           2.296   
3                                           2.319   
4                                           2.290   

   albania_gross_domestic_product,_current_prices  \
0                                           5.759   
1                                           6.663   
2                                           7.280   
3                                           7.649   
4                                           8.083   

   albania_gross_domestic_product,_deflator  \
0                                     5.935   
1                                     5.809   
2                                     5.814   
3                                     5.809   
4                                     5.623   

   albania_gross_domestic_product_per_capita,_constant_prices  \
0                                          116584.54            
1                                          120786.15            
2                                          121689.95            
3                                          120446.72            
4                                          120297.85            

   albania_gross_domestic_product_per_capita,_constant_prices  \
0                                            5557.56            
1                                            5757.84            
2                                            5800.93            
3                                            5741.66            
4                                            5734.57            

   albania_gross_domestic_product_per_capita,_current_prices  ...  \
0                                            6919.41          ...   
1                                            7016.15          ...   
2                                            7074.57          ...   
3                                            6997.24          ...   
4                                            6763.89          ...   

   sub-saharan_africa_direct_investment,_net  \
0                                      0.390   
1                                     -0.524   
2                                     -1.243   
3                                     -0.884   
4                                     -0.549   

   sub-saharan_africa_portfolio_investment,_net  \
0                                         0.706   
1                                         0.197   
2                                        -0.558   
3                                         0.627   
4                                        -1.062   

   sub-saharan_africa_other_investment,_net  \
0                                    -6.298   
1                                    -9.892   
2                                   -11.627   
3                                    -9.151   
4                                    -1.789   

 

In [688]:
# Now I can label the years
imf['year'] = imf['year'].astype(np.number)

crisis_years = [1975, 1982, 1991, 2009, 2020]
recovery_years = [1976, 1977, 1983, 1984, 1992, 1993, 2010, 2011]
downturn_years = [1958, 1998, 2001, 2012]

labels = []
for i in range(len(imf)):
    if imf['year'][i] in crisis_years:
        labels.append('recession')

    elif imf.loc[i]['year'] in recovery_years:
        labels.append('recovery')

    elif imf.loc[i]['year'] in downturn_years:
        labels.append('downturn')
        
    else:
        labels.append('expansion')

imf['year_label'] = labels

In [ ]:
imf[['year', 'year_label']]

In [ ]:
imf

### Model

In [ ]:
X = imf.drop(['year', 'year_label'],axis=1)

In [ ]:
Y = imf['year_label']

encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = pd.DataFrame(encoder.transform(Y))

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, encoded_Y, test_size=0.3, shuffle=None)

scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
y_test

In [ ]:
y_train

In [ ]:
# Build the model.
model = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(3724,)),
    Dense(64, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
    Dense(32, activation='sigmoid'),
    Dense(16, activation='sigmoid'),
    Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history = model.fit(
  X_train,
  to_categorical(y_train), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=10,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model.evaluate(
  X_test,
  to_categorical(y_test)
)

In [ ]:
predictions = model.predict(X_test)
print(np.argmax(predictions, axis=1))

predictions

In [ ]:
# The model accuracy is acceptable, however the high accuracy is only achieved because of the imbalanced data.
# The model is predicting every year as expansion
# I will try up- and downsampling

In [ ]:
all_data = imf
all_data

### Downsampling

In [ ]:
other_categories = all_data[all_data['year_label']!='expansion']
other_categories['year_label'].value_counts()

In [ ]:
category_expansion = all_data[all_data['year_label']=='expansion']
category_expansion_undersampled = resample(category_expansion, replace=False, n_samples = 6)
category_expansion_undersampled

In [ ]:
imf_undersampled = pd.concat([other_categories, category_expansion_undersampled], axis=0)
imf_undersampled.shape

In [ ]:
X_undersampled = imf_undersampled.drop(['year', 'year_label'],axis=1)

Y_undersampled = imf_undersampled['year_label']

encoder = LabelEncoder()
encoder.fit(Y_undersampled)
encoded_Y_undersampled = pd.DataFrame(encoder.transform(Y_undersampled))

X_train_under, X_test_under, y_train_under, y_test_under = train_test_split(X_undersampled, encoded_Y_undersampled,
                                                                            test_size=0.3)


scaler = preprocessing.StandardScaler()
X_train_under = scaler.fit_transform(X_train_under)
X_test_under = scaler.transform(X_test_under)

In [ ]:
# Build the model.
model_under = Sequential([ # as far as we know, all networks are sequential
    Dense(32, activation='sigmoid', input_shape=(3724,)),
    Dense(32, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model_under.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history_under = model_under.fit(
  X_train_under,
  to_categorical(y_train_under), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=20,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model_under.evaluate(
  X_test_under,
  to_categorical(y_test_under)
)

In [ ]:
predictions = model_under.predict(X_test_under)
print(np.argmax(predictions, axis=1))

predictions

In [ ]:
y_test_under

In [ ]:
# Much better!

## Upsampling

In [ ]:
category_expansion = all_data[all_data['year_label']=='expansion']
category_recession = all_data[all_data['year_label']=='recession']
category_recovery = all_data[all_data['year_label']=='recovery']
category_downturn = all_data[all_data['year_label']=='downturn']

category_recession_oversampled = resample(category_recession,replace=True, n_samples = len(category_expansion))
category_recovery_oversampled = resample(category_recovery,replace=True, n_samples = len(category_expansion))
category_downturn_oversampled = resample(category_downturn,replace=True, n_samples = len(category_expansion))

imf_upsampled= pd.concat([category_expansion,
                          category_recession_oversampled,
                          category_recovery_oversampled,
                          category_downturn_oversampled], axis=0)
imf_upsampled.shape

In [ ]:
X_updersampled = imf_upsampled.drop(['year', 'year_label'],axis=1)

Y_upsampled = imf_upsampled['year_label']

encoder = LabelEncoder()
encoder.fit(Y_upsampled)
encoded_Y_upsampled = pd.DataFrame(encoder.transform(Y_upsampled))

X_train_up, X_test_up, y_train_up, y_test_up = train_test_split(X_updersampled, encoded_Y_upsampled,
                                                                            test_size=0.15)


scaler = preprocessing.StandardScaler()
X_train_under = scaler.fit_transform(X_train_up)
X_test_under = scaler.transform(X_test_up)

In [ ]:
# Build the model.
model_up = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(3724,)),
    Dense(64, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
    Dense(32, activation='sigmoid'),
    Dense(16, activation='sigmoid'),
    Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model_up.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history_up = model_up.fit(
  X_train_up,
  to_categorical(y_train_up), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=20,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model_up.evaluate(
  X_test_up,
  to_categorical(y_test_up)
)

In [ ]:
predictions = model_up.predict(X_test_up)
print(np.argmax(predictions, axis=1))

predictions

In [ ]:
y_test_up

### Conclusions

None of the models was able to predict any of the recession years. Even with upsampling, almost every other category is being missed.

# World Bank Data

In [561]:
wb_gdp = pd.read_csv(r'C:\Users\luana\Ironhack DA\Unit 9\final_bootcamp_project\csv_files\wb_gdp_predictions.csv')
wb_gdp

WLD_gdp_per_capita_us$  WLD_gdp_per_capita_%_growth  \
0               471.544201                     2.457848   
1               496.165533                     3.531222   
2               523.989741                     3.038246   
3               562.186364                     4.414647   
4               600.107959                     3.423587   
5               637.822013                     3.529408   
6               665.055206                     2.064721   
7               703.361657                     3.807017   
8               759.933925                     3.651471   
9               813.897741                     1.799655   
10              880.266360                     2.096627   
11              994.694220                     3.516280   
12             1189.946669                     4.337687   
13             1345.558816                    -0.133413   
14             1471.502797                    -1.197047   
15             1571.389471                     3.446159   
16             1746.680853                     2.308947   
17             2021.612228                     2.347914   
18             2307.362289                     2.370826   
19             2556.954920                     0.126361   
20             2599.297941                     0.166623   
21             2527.674030                    -1.381191   
22             2532.623191                     0.850482   
23             2579.857213                     2.878301   
24             2657.496341                     1.917044   
25             3087.531735                     1.646270   
26             3452.776816                     1.914377   
27             3790.807087                     2.820068   
28             3890.962053                     1.980602   
29             4314.338715                     1.113071   
30             4426.034577                    -0.204415   
31             4659.467672                     0.489461   
32             4662.786795                     0.239882   
33             4957.636478                     1.759360   
34             5439.808718                     1.561335   
35             5481.614893                     2.137004   
36             5384.692144                     2.442608   
37             5297.268208                     1.383418   
38             5425.049261                     2.142579   
39             5533.052968                     3.122922   
40             5427.391189                     0.690283   
41             5565.592514                     1.039166   
42             6163.078955                     1.865792   
43             6859.583985                     3.199133   
44             7337.489822                     2.761759   
45             7854.104396                     3.192811   
46             8743.396033                     3.202076   
47             9489.966012                     0.815397   
48             8890.801150                    -2.516827   
49             9621.128288                     3.289390   
50            10544.875810                     2.112731   
51            10648.237225                     1.468547   
52            10815.580013                     1.583996   
53            10976.383454                     1.830545   
54            10231.702585                     1.876910   
55            10286.411330                     1.610004   
56            10825.902369                     2.200971   
57            11366.073669                     2.144471   
58            11407.479334                     1.531947   
59            10936.057466                    -4.269491   
60            11169.974582                    -0.334936   
61            11406.017489                     1.029786   
62            11644.205505                     1.503147   

    ARG_gdp_per_capita_us$  ARG_gdp_per_capita_%_growth  \
0              1184.627932                     3.728779   
1              1155.890717                    -2.425843   
2               850.304579                    -6.7

In [562]:
wb_gdp ['time'] = wb_gdp ['time'].str.replace('YR', '')

In [563]:
# Dropping 2021,2022 and 2023. I will label them later
wb_gdp ['time'] = wb_gdp ['time'].str.replace('YR', '')
wb_gdp_to_predict = wb_gdp[-3:]
wb_gdp = wb_gdp[:-3]
wb_gdp

WLD_gdp_per_capita_us$  WLD_gdp_per_capita_%_growth  \
0               471.544201                     2.457848   
1               496.165533                     3.531222   
2               523.989741                     3.038246   
3               562.186364                     4.414647   
4               600.107959                     3.423587   
5               637.822013                     3.529408   
6               665.055206                     2.064721   
7               703.361657                     3.807017   
8               759.933925                     3.651471   
9               813.897741                     1.799655   
10              880.266360                     2.096627   
11              994.694220                     3.516280   
12             1189.946669                     4.337687   
13             1345.558816                    -0.133413   
14             1471.502797                    -1.197047   
15             1571.389471                     3.446159   
16             1746.680853                     2.308947   
17             2021.612228                     2.347914   
18             2307.362289                     2.370826   
19             2556.954920                     0.126361   
20             2599.297941                     0.166623   
21             2527.674030                    -1.381191   
22             2532.623191                     0.850482   
23             2579.857213                     2.878301   
24             2657.496341                     1.917044   
25             3087.531735                     1.646270   
26             3452.776816                     1.914377   
27             3790.807087                     2.820068   
28             3890.962053                     1.980602   
29             4314.338715                     1.113071   
30             4426.034577                    -0.204415   
31             4659.467672                     0.489461   
32             4662.786795                     0.239882   
33             4957.636478                     1.759360   
34             5439.808718                     1.561335   
35             5481.614893                     2.137004   
36             5384.692144                     2.442608   
37             5297.268208                     1.383418   
38             5425.049261                     2.142579   
39             5533.052968                     3.122922   
40             5427.391189                     0.690283   
41             5565.592514                     1.039166   
42             6163.078955                     1.865792   
43             6859.583985                     3.199133   
44             7337.489822                     2.761759   
45             7854.104396                     3.192811   
46             8743.396033                     3.202076   
47             9489.966012                     0.815397   
48             8890.801150                    -2.516827   
49             9621.128288                     3.289390   
50            10544.875810                     2.112731   
51            10648.237225                     1.468547   
52            10815.580013                     1.583996   
53            10976.383454                     1.830545   
54            10231.702585                     1.876910   
55            10286.411330                     1.610004   
56            10825.902369                     2.200971   
57            11366.073669                     2.144471   
58            11407.479334                     1.531947   
59            10936.057466                    -4.269491   

    ARG_gdp_per_capita_us$  ARG_gdp_per_capita_%_growth  \
0              1184.627932                     3.728779   
1              1155.890717                    -2.425843   
2               850.304579                    -6.788240   
3              1173.238105                     8.437244   
4              1279.113778                     8.896809   
5              1272.803204                    -2.1

In [564]:
# Label the years

wb_gdp ['time'] = wb_gdp ['time'].str.replace('YR', '')


crisis_years = ['1975', '1982', '1991', '2009', '2020']
recovery_years = ['1976', '1977', '1983', '1984', '1992', '1993', '2010', '2011']
downturn_years = ['1958', '1998', '2001', '2012']

labels = []
for i in range(len(wb_gdp)):
    if wb_gdp ['time'][i] in crisis_years:
        labels.append('recession')

    elif wb_gdp['time'][i] in recovery_years:
        labels.append('recovery')

    elif wb_gdp ['time'][i] in downturn_years:
        labels.append('downturn')
        
    else:
        labels.append('expansion')

wb_gdp['year_label'] = labels

C:\Users\luana\AppData\Local\Temp\ipykernel_21664\141345650.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wb_gdp ['time'] = wb_gdp ['time'].str.replace('YR', '')
C:\Users\luana\AppData\Local\Temp\ipykernel_21664\141345650.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wb_gdp['year_label'] = labels


In [565]:
X = wb_gdp.drop(['time', 'year_label'],axis=1)

Y = wb_gdp['year_label']

encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = pd.DataFrame(encoder.transform(Y))
X_train, X_test, y_train, y_test = train_test_split(X, encoded_Y, test_size=0.3, shuffle=None)

scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(X.shape)

(60, 40)


In [566]:
# Build the model.
model = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(40,)),
    Dense(64, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
    Dense(32, activation='sigmoid'),
    Dense(16, activation='sigmoid'),
    Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history = model.fit(
  X_train,
  to_categorical(y_train), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=20,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model.evaluate(
  X_test,
  to_categorical(y_test)
)

Epoch 1/20
7/7 [==============================] - 1s 2ms/step - loss: 1.1298 - accuracy: 0.7381
Epoch 2/20
7/7 [==============================] - 0s 3ms/step - loss: 1.0560 - accuracy: 0.7381
Epoch 3/20
7/7 [==============================] - 0s 2ms/step - loss: 0.9949 - accuracy: 0.7381
Epoch 4/20
7/7 [==============================] - 0s 3ms/step - loss: 0.9541 - accuracy: 0.7381
Epoch 5/20
7/7 [==============================] - 0s 3ms/step - loss: 0.9300 - accuracy: 0.7381
Epoch 6/20
7/7 [==============================] - 0s 3ms/step - loss: 0.9055 - accuracy: 0.7381
Epoch 7/20
7/7 [==============================] - 0s 2ms/step - loss: 0.8954 - accuracy: 0.7381
Epoch 8/20
7/7 [==============================] - 0s 3ms/step - loss: 0.8783 - accuracy: 0.7381
Epoch 9/20
7/7 [==============================] - 0s 3ms/step - loss: 0.8735 - accuracy: 0.7381
Epoch 10/20
7/7 [==============================] - 0s 2ms/step - loss: 0.8648 - accuracy: 0.7381
Epoch 11/20
7/7 [======================

[0.8894979953765869, 0.7222222089767456]

In [567]:
predictions = model.predict(X_test)
print(np.argmax(predictions, axis=1))

predictions

1/1 [==============================] - 0s 115ms/step
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


array([[0.06654615, 0.711251  , 0.08572263, 0.13648023],
       [0.06642244, 0.7111225 , 0.08580007, 0.13665502],
       [0.0664174 , 0.71134126, 0.0856978 , 0.13654351],
       [0.06638524, 0.7117159 , 0.08551022, 0.13638866],
       [0.06633606, 0.71054554, 0.08607472, 0.1370437 ],
       [0.06640004, 0.71084595, 0.08593155, 0.13682245],
       [0.06640695, 0.7109955 , 0.08586323, 0.13673432],
       [0.06632764, 0.71038276, 0.08615166, 0.13713793],
       [0.0662938 , 0.7097566 , 0.08644325, 0.13750629],
       [0.06647029, 0.71137875, 0.08567846, 0.13647251],
       [0.06634283, 0.7100383 , 0.08630884, 0.13730998],
       [0.06642808, 0.7118866 , 0.08542362, 0.13626175],
       [0.06655991, 0.7115682 , 0.08557009, 0.13630182],
       [0.06645274, 0.7109825 , 0.08587421, 0.13669053],
       [0.06643219, 0.71177804, 0.08548436, 0.13630542],
       [0.06646612, 0.70877635, 0.08688901, 0.13786845],
       [0.06655757, 0.7116024 , 0.08555486, 0.13628514],
       [0.06653945, 0.7112392 ,

In [568]:
# Again, it can only predict the expansion years

In [569]:
columns_percentage = ['time', 'year_label']

for column in wb_gdp.columns:
    if '%' in column:
        columns_percentage.append(column)

        
columns_percentage

['time',
 'year_label',
 'WLD_gdp_per_capita_%_growth',
 'ARG_gdp_per_capita_%_growth',
 'AUS_gdp_per_capita_%_growth',
 'BRA_gdp_per_capita_%_growth',
 'CAN_gdp_per_capita_%_growth',
 'CHN_gdp_per_capita_%_growth',
 'FRA_gdp_per_capita_%_growth',
 'DEU_gdp_per_capita_%_growth',
 'IND_gdp_per_capita_%_growth',
 'IDN_gdp_per_capita_%_growth',
 'ITA_gdp_per_capita_%_growth',
 'JPN_gdp_per_capita_%_growth',
 'MEX_gdp_per_capita_%_growth',
 'RUS_gdp_per_capita_%_growth',
 'SAU_gdp_per_capita_%_growth',
 'ZAF_gdp_per_capita_%_growth',
 'KOR_gdp_per_capita_%_growth',
 'TUR_gdp_per_capita_%_growth',
 'GBR_gdp_per_capita_%_growth',
 'USA_gdp_per_capita_%_growth']

In [570]:
gdp_percentage = wb_gdp[columns_percentage]
gdp_percentage

time year_label  WLD_gdp_per_capita_%_growth  ARG_gdp_per_capita_%_growth  \
0   1961  expansion                     2.457848                     3.728779   
1   1962  expansion                     3.531222                    -2.425843   
2   1963  expansion                     3.038246                    -6.788240   
3   1964  expansion                     4.414647                     8.437244   
4   1965  expansion                     3.423587                     8.896809   
5   1966  expansion                     3.529408                    -2.136478   
6   1967  expansion                     2.064721                     1.678435   
7   1968  expansion                     3.807017                     3.286938   
8   1969  expansion                     3.651471                     8.050279   
9   1970  expansion                     1.799655                     1.479494   
10  1971  expansion                     2.096627                     4.007465   
11  1972  expansion                     3.516280                     0.005836   
12  1973  expansion                     4.337687                     1.156921   
13  1974  expansion                    -0.133413                     3.851758   
14  1975  recession                    -1.197047                    -1.587829   
15  1976   recovery                     3.446159                    -3.506391   
16  1977   recovery                     2.308947                     5.342990   
17  1978  expansion                     2.347914                    -5.916401   
18  1979  expansion                     2.370826                     8.577690   
19  1980  expansion                     0.126361                    -0.029610   
20  1981  expansion                     0.166623                    -6.668498   
21  1982  recession                    -1.381191                    -2.307766   
22  1983   recovery                     0.850482                     2.681982   
23  1984   recovery                     2.878301                    -0.051847   
24  1985  expansion                     1.917044                    -6.692591   
25  1986  expansion                     1.646270                     4.484327   
26  1987  expansion                     1.914377                     1.106065   
27  1988  expansion                     2.820068                    -2.603236   
28  1989  expansion                     1.980602                    -8.540862   
29  1990  expansion                     1.113071                    -3.874085   
30  1991  recession                    -0.204415                     7.614326   
31  1992   recovery                     0.489461                     6.487632   
32  1993   recovery                     0.239882                     6.802927   
33  1994  expansion                     1.759360                     4.505457   
34  1995  expansion                     1.561335                    -4.032159   
35  1996  expansion                     2.137004                     4.274586   
36  1997  expansion                     2.442608                     6.864733   
37  1998   downturn                     1.383418                     2.681003   
38  1999  expansion                     2.142579                    -4.455178   
39  2000  expansion                     3.122922                    -1.874936   
40  2001   downturn                     0.690283                    -5.447048   
41  2002  expansion                     1.039166                   -11.854782   
42  2003  expansion                     1.865792                     7.676534   
43  2004  expansion                     3.199133                     7.884938   
44  2005  expansion                     2.761759                     7.729489   
45  2006  expansion                     3.192811                     6.955534   
46  2007  expansion                     3.202076                     7.924211   
47  2008  expansion                     0.815397                     3.029501   
48  2009  recession                

In [571]:
X = gdp_percentage.drop(['time', 'year_label'],axis=1)

Y = gdp_percentage['year_label']

encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = pd.DataFrame(encoder.transform(Y))
X_train, X_test, y_train, y_test = train_test_split(X, encoded_Y, test_size=0.35, shuffle=None)

scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(X.shape)

(60, 20)


In [572]:
# Build the model.
model = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(20,)),
    Dense(128, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
#     Dense(32, activation='sigmoid'),
#     Dense(16, activation='sigmoid'),
#     Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history = model.fit(
  X_train,
  to_categorical(y_train), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=20,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model.evaluate(
  X_test,
  to_categorical(y_test)
)

Epoch 1/20
7/7 [==============================] - 1s 2ms/step - loss: 1.2580 - accuracy: 0.4872
Epoch 2/20
7/7 [==============================] - 0s 3ms/step - loss: 0.9548 - accuracy: 0.6667
Epoch 3/20
7/7 [==============================] - 0s 2ms/step - loss: 0.9719 - accuracy: 0.6667
Epoch 4/20
7/7 [==============================] - 0s 2ms/step - loss: 0.9137 - accuracy: 0.6667
Epoch 5/20
7/7 [==============================] - 0s 2ms/step - loss: 0.8935 - accuracy: 0.6667
Epoch 6/20
7/7 [==============================] - 0s 2ms/step - loss: 0.8651 - accuracy: 0.6667
Epoch 7/20
7/7 [==============================] - 0s 2ms/step - loss: 0.8435 - accuracy: 0.6667
Epoch 8/20
7/7 [==============================] - 0s 2ms/step - loss: 0.8231 - accuracy: 0.7179
Epoch 9/20
7/7 [==============================] - 0s 3ms/step - loss: 0.7965 - accuracy: 0.7179
Epoch 10/20
7/7 [==============================] - 0s 2ms/step - loss: 0.7722 - accuracy: 0.7179
Epoch 11/20
7/7 [======================

[0.5681012272834778, 0.8571428656578064]

In [573]:
predictions = model.predict(X_test)
print(np.argmax(predictions, axis=1))

predictions

1/1 [==============================] - 0s 67ms/step
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


array([[0.03369792, 0.85397696, 0.01222778, 0.10009731],
       [0.07253859, 0.5277977 , 0.08827052, 0.3113932 ],
       [0.08760554, 0.5767098 , 0.1844024 , 0.15128224],
       [0.03519737, 0.845736  , 0.01496629, 0.1041003 ],
       [0.03657202, 0.7478079 , 0.02180177, 0.19381823],
       [0.03139187, 0.83028996, 0.01045119, 0.12786694],
       [0.06262567, 0.51145554, 0.08750664, 0.3384122 ],
       [0.04941919, 0.7554308 , 0.02714007, 0.16800989],
       [0.04451082, 0.7752033 , 0.0306844 , 0.1496015 ],
       [0.06371427, 0.6797378 , 0.06055267, 0.19599526],
       [0.02554888, 0.9023786 , 0.00696037, 0.0651122 ],
       [0.02622475, 0.8855088 , 0.00567699, 0.08258945],
       [0.03748483, 0.85600215, 0.01205161, 0.09446139],
       [0.04753204, 0.76656044, 0.03072873, 0.1551788 ],
       [0.07784919, 0.512681  , 0.0969847 , 0.31248507],
       [0.03903766, 0.7970745 , 0.02003297, 0.14385487],
       [0.06714635, 0.65822035, 0.09402805, 0.1806053 ],
       [0.04509513, 0.72049856,

In [574]:
y_test

0
43  1
33  1
40  0
45  1
5   1
27  1
34  1
4   1
28  1
56  1
9   1
3   1
2   1
50  3
24  1
49  3
41  1
17  1
46  1
25  1
6   1

In [597]:
# Removing the columns in dolar seems to have improved the predictions
# I will upsample again and downsample again

category_expansion = gdp_percentage[gdp_percentage['year_label']=='expansion']
category_recession = gdp_percentage[gdp_percentage['year_label']=='recession']
category_recovery = gdp_percentage[gdp_percentage['year_label']=='recovery']
category_downturn = gdp_percentage[gdp_percentage['year_label']=='downturn']

category_recession_oversampled = resample(category_recession,replace=True, n_samples = len(category_expansion))
category_recovery_oversampled = resample(category_recovery,replace=True, n_samples = len(category_expansion))
category_downturn_oversampled = resample(category_downturn,replace=True, n_samples = len(category_expansion))

gdp_percentage_upsampled = pd.concat([category_expansion,
                          category_recession_oversampled,
                          category_recovery_oversampled,
                          category_downturn_oversampled], axis=0)

gdp_percentage_upsampled.shape

(176, 22)

In [623]:
X = gdp_percentage_upsampled.drop(['time', 'year_label'],axis=1)

Y = gdp_percentage_upsampled['year_label']

encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = pd.DataFrame(encoder.transform(Y))
X_train, X_test, y_train, y_test = train_test_split(X, encoded_Y, test_size=0.35, shuffle=None)

scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(X.shape)

(176, 20)


In [628]:
# 1: Expansion, 2: recession, 3: recovery, 0: downturn

In [599]:
# Build the model.
model = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(20,)),
    Dense(128, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
#     Dense(32, activation='sigmoid'),
#     Dense(16, activation='sigmoid'),
#     Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history = model.fit(
  X_train,
  to_categorical(y_train), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=20,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model.evaluate(
  X_test,
  to_categorical(y_test)
)

Epoch 1/20
19/19 [==============================] - 1s 2ms/step - loss: 1.4360 - accuracy: 0.2895
Epoch 2/20
19/19 [==============================] - 0s 2ms/step - loss: 1.3001 - accuracy: 0.4912
Epoch 3/20
19/19 [==============================] - 0s 2ms/step - loss: 1.2094 - accuracy: 0.4737
Epoch 4/20
19/19 [==============================] - 0s 2ms/step - loss: 1.1068 - accuracy: 0.5789
Epoch 5/20
19/19 [==============================] - 0s 2ms/step - loss: 1.0059 - accuracy: 0.7018
Epoch 6/20
19/19 [==============================] - 0s 2ms/step - loss: 0.9135 - accuracy: 0.7193
Epoch 7/20
19/19 [==============================] - 0s 2ms/step - loss: 0.8342 - accuracy: 0.7544
Epoch 8/20
19/19 [==============================] - 0s 2ms/step - loss: 0.7610 - accuracy: 0.7368
Epoch 9/20
19/19 [==============================] - 0s 2ms/step - loss: 0.6855 - accuracy: 0.7193
Epoch 10/20
19/19 [==============================] - 0s 2ms/step - loss: 0.6215 - accuracy: 0.8158
Epoch 11/20
19/19 [

[0.2806028127670288, 0.9354838728904724]

In [600]:
predictions = model.predict(X_test)
print(np.argmax(predictions, axis=1))

2/2 [==============================] - 0s 2ms/step
[2 3 2 2 2 3 3 1 1 3 2 0 3 0 2 0 3 0 0 0 2 1 2 0 0 2 3 3 2 1 0 2 0 3 0 3 3
 0 1 0 1 1 3 2 3 3 0 3 0 3 1 3 2 1 3 3 3 0 3 2 3 2]


In [617]:
predictions_array =np.argmax(predictions, axis=1)
y_test_df = pd.DataFrame(y_test).reset_index(drop=True)
y_test_df.columns= ['y_test']
predictions_df = pd.DataFrame(predictions_array)
predictions_df.columns= ['predicted_y']
predictions_df = pd.concat([predictions_df,y_test_df], axis=1)
pd.set_option('display.max_rows',None)
predictions_df

predicted_y  y_test
0             2       2
1             3       3
2             2       2
3             2       2
4             2       2
5             3       3
6             3       3
7             1       1
8             1       1
9             3       3
10            2       2
11            0       0
12            3       3
13            0       0
14            2       2
15            0       0
16            3       3
17            0       0
18            0       0
19            0       0
20            2       2
21            1       1
22            2       2
23            0       0
24            0       0
25            2       2
26            3       3
27            3       3
28            2       2
29            1       1
30            0       0
31            2       2
32            0       1
33            3       3
34            0       0
35            3       1
36            3       3
37            0       0
38            1       1
39            0       0
40            1       1
41            1       1
42            3       3
43            2       2
44            3       1
45            3       3
46            0       0
47            3       3
48            0       0
49            3       3
50            1       1
51            3       3
52            2       2
53            1       1
54            3       3
55            3       1
56            3       3
57            0       0
58            3       3
59            2       2
60            3       3
61            2       2

In [584]:
# This model has a much better accuracy! Will be used for predictions
model_upsampled = model 
scaler_upsampled  = scaler

filename = 'model_gdp_upsampled.sav'
pickle.dump(model, open(filename, 'wb'))

filename= 'scaler_gdp_upsample.sav'
pickle.dump(scaler, open(filename, 'wb'))

INFO:tensorflow:Assets written to: C:\Users\luana\AppData\Local\Temp\tmpdemxf2sj\assets


In [585]:
# Just in case I will also downsample
gdp_percentage['year_label'].value_counts()

expansion    44
recovery      8
recession     5
downturn      3
Name: year_label, dtype: int64

In [586]:
category_expansion = gdp_percentage[gdp_percentage['year_label']=='expansion']
other_categories = gdp_percentage[gdp_percentage['year_label']!='expansion']
category_category_expansion_undersampled = resample(category_expansion,replace=True, n_samples = 8)

gdp_percentage_undersampled = pd.concat([other_categories, category_category_expansion_undersampled], axis=0)

gdp_percentage_undersampled.shape

(24, 22)

In [587]:
X = gdp_percentage_undersampled.drop(['time', 'year_label'],axis=1)

Y = gdp_percentage_undersampled['year_label']

encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = pd.DataFrame(encoder.transform(Y))
X_train, X_test, y_train, y_test = train_test_split(X, encoded_Y, test_size=0.15, shuffle=None)

scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(X.shape)

(24, 20)


In [588]:
# Build the model.
model = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(20,)),
    Dense(128, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
    Dense(32, activation='sigmoid'),
    Dense(16, activation='sigmoid'),
    Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history = model.fit(
  X_train,
  to_categorical(y_train), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=20,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model.evaluate(
  X_test,
  to_categorical(y_test)
)

Epoch 1/20
4/4 [==============================] - 1s 3ms/step - loss: 1.3425 - accuracy: 0.3000
Epoch 2/20
4/4 [==============================] - 0s 3ms/step - loss: 1.3249 - accuracy: 0.3000
Epoch 3/20
4/4 [==============================] - 0s 4ms/step - loss: 1.3080 - accuracy: 0.3000
Epoch 4/20
4/4 [==============================] - 0s 4ms/step - loss: 1.3007 - accuracy: 0.3000
Epoch 5/20
4/4 [==============================] - 0s 3ms/step - loss: 1.2926 - accuracy: 0.3000
Epoch 6/20
4/4 [==============================] - 0s 3ms/step - loss: 1.2901 - accuracy: 0.3000
Epoch 7/20
4/4 [==============================] - 0s 4ms/step - loss: 1.2883 - accuracy: 0.3000
Epoch 8/20
4/4 [==============================] - 0s 4ms/step - loss: 1.2875 - accuracy: 0.2500
Epoch 9/20
4/4 [==============================] - 0s 3ms/step - loss: 1.2851 - accuracy: 0.4000
Epoch 10/20
4/4 [==============================] - 0s 3ms/step - loss: 1.2858 - accuracy: 0.4000
Epoch 11/20
4/4 [======================

[1.5776002407073975, 0.0]

In [589]:
predictions = model.predict(X_test)
print(np.argmax(predictions, axis=1))

1/1 [==============================] - 0s 102ms/step
[1 1 1 1]


In [590]:
y_test

0
4   3
10  0
1   3
3   2

In [591]:
y_train

0
9   0
11  2
14  0
15  2
12  3
20  1
22  1
7   3
21  1
5   3
6   2
16  1
18  1
13  3
2   3
19  1
23  1
8   3
0   2
17  1

In [592]:
# Not great... moving on with the upsampled model

In [593]:
wb_gdp_to_predict

WLD_gdp_per_capita_us$  WLD_gdp_per_capita_%_growth  \
60            11169.974582                    -0.334936   
61            11406.017489                     1.029786   
62            11644.205505                     1.503147   

    ARG_gdp_per_capita_us$  ARG_gdp_per_capita_%_growth  \
60             8570.634519                    -0.991484   
61             8556.235071                     0.504945   
62             8542.467406                     0.734062   

    AUS_gdp_per_capita_us$  AUS_gdp_per_capita_%_growth  \
60            52389.355728                     1.224230   
61            53095.141956                     1.708921   
62            53797.690131                     1.802788   

    BRA_gdp_per_capita_us$  BRA_gdp_per_capita_%_growth  \
60             6837.643155                    -1.241252   
61             6859.729819                     0.342835   
62             6881.155986                     1.098611   

    CAN_gdp_per_capita_us$  CAN_gdp_per_capita_%_growth  ...  \
60            43804.966920                    -2.114828  ...   
61            44347.911525                     0.382067  ...   
62            44887.123525                     1.864915  ...   

    ZAF_gdp_per_capita_%_growth  KOR_gdp_per_capita_us$  \
60                    -4.075358            32529.782043   
61                    -2.083890            33479.340388   
62                    -0.963730            34446.500011   

    KOR_gdp_per_capita_%_growth  TUR_gdp_per_capita_us$  \
60                     3.344573             8625.572010   
61                     4.958773             8713.704291   
62                     5.560093             8800.841526   

    TUR_gdp_per_capita_%_growth  GBR_gdp_per_capita_us$  \
60                     2.785851            41604.094252   
61                     2.786096            42106.029471   
62                     2.786097            42603.917211   

    GBR_gdp_per_capita_%_growth  USA_gdp_per_capita_us$  \
60                    -2.305666            64639.616426   
61                     0.284902            66277.833117   
62                     1.204767            67942.758047   

    USA_gdp_per_capita_%_growth  time  
60                     0.129421  2021  
61                     1.385264  2022  
62                     1.739038  2023  

[3 rows x 41 columns]

In [594]:
X = wb_gdp_to_predict.drop('time', axis=1)
X = wb_gdp_to_predict.drop([column for column in wb_gdp_to_predict.columns if '%' not in column], axis=1)
X

WLD_gdp_per_capita_%_growth  ARG_gdp_per_capita_%_growth  \
60                    -0.334936                    -0.991484   
61                     1.029786                     0.504945   
62                     1.503147                     0.734062   

    AUS_gdp_per_capita_%_growth  BRA_gdp_per_capita_%_growth  \
60                     1.224230                    -1.241252   
61                     1.708921                     0.342835   
62                     1.802788                     1.098611   

    CAN_gdp_per_capita_%_growth  CHN_gdp_per_capita_%_growth  \
60                    -2.114828                     5.462254   
61                     0.382067                     6.773980   
62                     1.864915                     7.270462   

    FRA_gdp_per_capita_%_growth  DEU_gdp_per_capita_%_growth  \
60                    -4.279322                     0.556327   
61                    -1.980928                     1.485164   
62                    -0.574128                     1.650952   

    IND_gdp_per_capita_%_growth  IDN_gdp_per_capita_%_growth  \
60                     1.199434                     0.988048   
61                     2.744545                     2.383816   
62                     3.018479                     2.860059   

    ITA_gdp_per_capita_%_growth  JPN_gdp_per_capita_%_growth  \
60                    -3.443038                    -1.697046   
61                    -0.907248                    -0.134418   
62                     0.340169                     0.831293   

    MEX_gdp_per_capita_%_growth  RUS_gdp_per_capita_%_growth  \
60                    -1.676989                    -1.357894   
61                     0.519137                    -0.633264   
62                     1.165903                    -0.165620   

    SAU_gdp_per_capita_%_growth  ZAF_gdp_per_capita_%_growth  \
60                    -0.912741                    -4.075358   
61                     0.508637                    -2.083890   
62                     0.936077                    -0.963730   

    KOR_gdp_per_capita_%_growth  TUR_gdp_per_capita_%_growth  \
60                     3.344573                     2.785851   
61                     4.958773                     2.786096   
62                     5.560093                     2.786097   

    GBR_gdp_per_capita_%_growth  USA_gdp_per_capita_%_growth  
60                    -2.305666                     0.129421  
61                     0.284902                     1.385264  
62                     1.204767                     1.739038

In [595]:
# load the model from disk

filename = 'model_gdp_upsampled.sav'
loaded_model = pickle.load(open(filename, 'rb'))

filename= 'scaler_gdp_upsample.sav'
loaded_scaler = pickle.load(open(filename, 'rb'))

In [596]:

X_scaled = loaded_scaler.transform(X)

predictions = loaded_model.predict(X_scaled)
print(np.argmax(predictions, axis=1))

1/1 [==============================] - 0s 64ms/step
[2 2 1]


In [ ]:
# 1: Expansion, 2: recession, 3: recovery, 0: downturn

# Conclusion:

#### This model has predicted recession for 2021 and 2022, but expansion for 2023

# World Bank: data from 1981

In [675]:
pd.set_option('display.max_rows', 10)
wb_1981 = pd.read_csv(r'C:\Users\luana\Ironhack DA\Unit 9\final_bootcamp_project\csv_files\wb_from_1981_predictions.csv')
wb_1981

time  WLD_gdp_per_capita_us$  WLD_gdp_per_capita_%_growth  \
0   YR1981             2599.297941                     0.166623   
1   YR1982             2527.674030                    -1.381191   
2   YR1983             2532.623191                     0.850482   
3   YR1984             2579.857213                     2.878301   
4   YR1985             2657.496341                     1.917044   
..     ...                     ...                          ...   
38  YR2019            11407.479334                     1.531947   
39  YR2020            10936.057466                    -4.269491   
40  YR2021            11127.421453                     0.046068   
41  YR2022            11317.806416                     1.111767   
42  YR2023            11507.217363                     1.374935   

    WLD_consumer_price_index_%  WLD_unemployment_%_of_total_labor_force  \
0                    12.471612                                 5.681880   
1                    10.240268                                 5.681880   
2                     8.771147                                 5.681880   
3                     8.116398                                 5.681880   
4                     6.856812                                 5.681880   
..                         ...                                      ...   
38                    2.186902                                 5.356915   
39                    1.920968                                 6.573234   
40                    2.440144                                 6.204328   
41                    2.843239                                 5.993507   
42                    3.156206                                 5.873028   

    WLD_government_expense_%_gdp  WLD_industry_value_added_us$  \
0                      26.023013                     26.023013   
1                      26.023013                     26.023013   
2                      26.023013                     26.023013   
3                      26.023013                     26.023013   
4                      26.023013                     26.023013   
..                           ...                           ...   
38                     26.956431                     26.956431   
39                     34.063097                     34.063097   
40                     33.670439                     33.670439   
41                     33.306294                     33.306294   
42                     32.968592                     32.968592   

    ARG_gdp_per_capita_us$  ARG_gdp_per_capita_%_growth  \
0              2776.322088                    -6.668498   
1              2927.897357                    -2.307766   
2              3553.377509                     2.681982   
3              2659.708242                    -0.051847   
4              2926.126485                    -6.692591   
..                     ...                          ...   
38            10076.355241                    -2.994388   
39             8585.694742                   -10.765108   
40             8619.597289                    -2.626874   
41             8650.472253                    -0.347035   
42             8678.590006                     0.291637   

    ARG_unemployment_%_of_total_labor_force  ...  USA_gdp_per_capita_%_growth  \
0                                 10.980400  ...                     1.536320   
1                                 10.980400  ...                    -2.734570   
2                                 10.980400  ...                     3.631979   
3                                 10.980400  ...                     6.312168   
4                                 10.980400  ...                     3.250656   
..                                      ...  ...                          ...   
38                                 9.840000  ...                     1.824124   
39                                11.460000  ...                    -4.328618   
40                                11.386141  ...                    -

In [676]:
wb_1981 ['time'] = wb_1981 ['time'].str.replace('YR', '')
percentage_columns_1981 = ['time']
for column in wb_1981.columns:
    if "%" in column:
        percentage_columns_1981.append(column)

wb_1981  = wb_1981[percentage_columns_1981]
wb_1981_to_predict = wb_1981[-3:]
wb_1981 = wb_1981[:-3]
wb_1981

time  WLD_gdp_per_capita_%_growth  WLD_consumer_price_index_%  \
0   1981                     0.166623                   12.471612   
1   1982                    -1.381191                   10.240268   
2   1983                     0.850482                    8.771147   
3   1984                     2.878301                    8.116398   
4   1985                     1.917044                    6.856812   
..   ...                          ...                         ...   
35  2016                     1.610004                    1.550016   
36  2017                     2.200971                    2.192010   
37  2018                     2.144471                    2.438737   
38  2019                     1.531947                    2.186902   
39  2020                    -4.269491                    1.920968   

    WLD_unemployment_%_of_total_labor_force  WLD_government_expense_%_gdp  \
0                                  5.681880                     26.023013   
1                                  5.681880                     26.023013   
2                                  5.681880                     26.023013   
3                                  5.681880                     26.023013   
4                                  5.681880                     26.023013   
..                                      ...                           ...   
35                                 5.656542                     26.537560   
36                                 5.557650                     26.433629   
37                                 5.389680                     26.296561   
38                                 5.356915                     26.956431   
39                                 6.573234                     34.063097   

    ARG_gdp_per_capita_%_growth  ARG_unemployment_%_of_total_labor_force  \
0                     -6.668498                                  10.9804   
1                     -2.307766                                  10.9804   
2                      2.681982                                  10.9804   
3                     -0.051847                                  10.9804   
4                     -6.692591                                  10.9804   
..                          ...                                      ...   
35                    -3.110064                                   8.1110   
36                     1.757648                                   8.3500   
37                    -3.601610                                   9.2200   
38                    -2.994388                                   9.8400   
39                   -10.765108                                  11.4600   

    ARG_industry_value_added_%_growth  AUS_gdp_per_capita_%_growth  \
0                          -10.663334                     1.711915   
1                           -3.209381                     1.614208   
2                            5.377688                    -3.437117   
3                            0.674286                     3.418345   
4                           -8.651012                     3.822749   
..                                ...                          ...   
35                          -5.825159                     1.149347   
36                           2.302833                     0.587161   
37                          -3.049315                     1.302841   
38                          -4.790141                     0.571103   
39                          -9.369053                    -1.278526   

    AUS_consumer_price_index_%  ...  GBR_consumer_price_index_%  \
0                     9.487666  ...                   11.876627   
1                    11.351820  ...                    8.598864   
2                    10.038911  ...                    4.609303   
3                     3.960396  ...                    4.960711   
4                     6.734694  ...                    6.071394   
..                         ...  ...                         ...   
35                    1.276

In [641]:
wb_1981_to_predict

time  WLD_gdp_per_capita_%_growth  WLD_consumer_price_index_%  \
40  2021                     0.046068                    2.440144   
41  2022                     1.111767                    2.843239   
42  2023                     1.374935                    3.156206   

    WLD_unemployment_%_of_total_labor_force  WLD_government_expense_%_gdp  \
40                                 6.204328                     33.670439   
41                                 5.993507                     33.306294   
42                                 5.873028                     32.968592   

    ARG_gdp_per_capita_%_growth  ARG_unemployment_%_of_total_labor_force  \
40                    -2.626874                                11.386141   
41                    -0.347035                                11.325219   
42                     0.291637                                11.274967   

    ARG_industry_value_added_%_growth  AUS_gdp_per_capita_%_growth  \
40                          -1.386956                     1.009692   
41                           0.512524                     1.470736   
42                           0.964538                     1.563630   

    AUS_consumer_price_index_%  ...  GBR_consumer_price_index_%  \
40                    1.353727  ...                    1.457070   
41                    1.739290  ...                    1.775478   
42                    2.032605  ...                    1.992303   

    GBR_unemployment_%_of_total_labor_force  \
40                                 4.580937   
41                                 4.681121   
42                                 4.773256   

    GBR_government_debt_total_%_of_gdp  GBR_government_expense_%_gdp  \
40                          191.566660                     46.260029   
41                          192.347959                     45.422538   
42                          193.116626                     44.682401   

    GBR_industry_value_added_%_growth  USA_gdp_per_capita_%_growth  \
40                          -2.003782                    -0.276751   
41                          -0.203435                     0.968701   
42                           0.164021                     1.351526   

    USA_consumer_price_index_%  USA_unemployment_%_of_total_labor_force  \
40                    1.925820                                 7.541917   
41                    2.246906                                 7.165148   
42                    2.395839                                 6.885755   

    USA_government_debt_total_%_of_gdp  USA_government_expense_%_gdp  
40                          176.148250                     32.960833  
41                          192.481725                     32.985316  
42                          210.859812                     33.009191  

[3 rows x 92 columns]

In [677]:
# Label the years

crisis_years = ['1975', '1982', '1991', '2009', '2020']
recovery_years = ['1976', '1977', '1983', '1984', '1992', '1993', '2010', '2011']
downturn_years = ['1958', '1998', '2001', '2012']

labels = []
for i in range(len(wb_1981)):
    if wb_1981 ['time'][i] in crisis_years:
        labels.append('recession')

    elif wb_1981['time'][i] in recovery_years:
        labels.append('recovery')

    elif wb_1981 ['time'][i] in downturn_years:
        labels.append('downturn')
        
    else:
        labels.append('expansion')

wb_1981['year_label'] = labels

In [678]:
category_expansion = wb_1981[wb_1981['year_label']=='expansion']
category_recession = wb_1981[wb_1981['year_label']=='recession']
category_recovery = wb_1981[wb_1981['year_label']=='recovery']
category_downturn = wb_1981[wb_1981['year_label']=='downturn']

category_recession_oversampled = resample(category_recession,replace=True, n_samples = len(category_expansion))
category_recovery_oversampled = resample(category_recovery,replace=True, n_samples = len(category_expansion))
category_downturn_oversampled = resample(category_downturn,replace=True, n_samples = len(category_expansion))

wb_1981_upsampled = pd.concat([category_expansion,
                          category_recession_oversampled,
                          category_recovery_oversampled,
                          category_downturn_oversampled], axis=0)

X = wb_1981_upsampled.drop(['time', 'year_label'],axis=1)

Y = wb_1981_upsampled['year_label']

encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = pd.DataFrame(encoder.transform(Y))
X_train, X_test, y_train, y_test = train_test_split(X, encoded_Y, test_size=0.35, shuffle=None)

scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(X.shape)

(108, 91)


In [ ]:
# 1: Expansion, 2: recession , 3:recovery , 4: downturn

In [667]:
# Build the model.
model = Sequential([ # as far as we know, all networks are sequential
    Dense(128, activation='sigmoid', input_shape=(91,)),
    Dense(64, activation='sigmoid'),      # choosing relu instead of sigmoid, this is somewhat common
    Dense(32, activation='sigmoid'),
    Dense(16, activation='sigmoid'),
    Dense(8, activation='sigmoid'), 
    Dense(4, activation='softmax'),   # the softmax actiavation is the last one to compensate for the high volume additions
])

# Compile the model.
model.compile(
  optimizer='adam', #here we could use stochastic gradient descent, but adam is a de facto standard
  loss='categorical_crossentropy', #this is how we create the original blame to play the blame game
  metrics=['accuracy'],
)

# Train the model.
history = model.fit(
  X_train,
  to_categorical(y_train), # just to make sure the outputs are not considered numeric (because, ya know, they are numbers...)
  epochs=50,
  batch_size=6, # send 32 images at a time before you tweak the network again, to make it faster
)
 
    # Evaluate the model.
model.evaluate(
  X_test,
  to_categorical(y_test)
)


Epoch 1/50
12/12 [==============================] - 1s 2ms/step - loss: 1.7856 - accuracy: 0.2571
Epoch 2/50
12/12 [==============================] - 0s 3ms/step - loss: 1.6784 - accuracy: 0.2571
Epoch 3/50
12/12 [==============================] - 0s 2ms/step - loss: 1.6009 - accuracy: 0.2571
Epoch 4/50
12/12 [==============================] - 0s 2ms/step - loss: 1.5440 - accuracy: 0.2571
Epoch 5/50
12/12 [==============================] - 0s 2ms/step - loss: 1.5046 - accuracy: 0.2571
Epoch 6/50
12/12 [==============================] - 0s 2ms/step - loss: 1.4798 - accuracy: 0.2571
Epoch 7/50
12/12 [==============================] - 0s 3ms/step - loss: 1.4597 - accuracy: 0.2571
Epoch 8/50
12/12 [==============================] - 0s 2ms/step - loss: 1.4453 - accuracy: 0.2571
Epoch 9/50
12/12 [==============================] - 0s 2ms/step - loss: 1.4308 - accuracy: 0.2571
Epoch 10/50
12/12 [==============================] - 0s 3ms/step - loss: 1.4207 - accuracy: 0.2571
Epoch 11/50
12/12 [

[0.863950252532959, 0.8157894611358643]

In [669]:
predictions = model.predict(X_test)
print(np.argmax(predictions, axis=1))

2/2 [==============================] - 0s 3ms/step
[0 0 1 2 0 2 2 3 0 1 1 0 0 2 3 2 2 3 3 3 0 1 2 0 3 2 0 2 0 3 3 0 3 0 2 1 0
 1]


In [666]:
y_test

0
84   0
5    1
9    1
53   2
102  0
42   2
35   2
80   3
104  0
24   1
8    1
106  0
90   0
19   1
73   3
36   2
34   2
54   3
56   3
69   3
21   1
17   1
0    1
26   1
57   3
30   2
11   1
40   2
94   0
67   3
71   3
87   0
55   3
85   0
31   2
12   1
6    1
16   1

In [670]:
# Also here the accuracy is quite good!


# This model has a much better accuracy! Will be used for predictions
model_1981_upsampled = model 
scaler_1981_upsampled  = scaler

filename = 'model_1981_upsampled.sav'
pickle.dump(model_1981_upsampled, open(filename, 'wb'))

filename= 'scaler_1981_upsampled.sav'
pickle.dump(scaler_1981_upsampled, open(filename, 'wb'))


INFO:tensorflow:Assets written to: C:\Users\luana\AppData\Local\Temp\tmpe53ejhh_\assets


In [672]:
X = wb_1981_to_predict.drop(['time'], axis=1)
X

WLD_gdp_per_capita_%_growth  WLD_consumer_price_index_%  \
40                     0.046068                    2.440144   
41                     1.111767                    2.843239   
42                     1.374935                    3.156206   

    WLD_unemployment_%_of_total_labor_force  WLD_government_expense_%_gdp  \
40                                 6.204328                     33.670439   
41                                 5.993507                     33.306294   
42                                 5.873028                     32.968592   

    ARG_gdp_per_capita_%_growth  ARG_unemployment_%_of_total_labor_force  \
40                    -2.626874                                11.386141   
41                    -0.347035                                11.325219   
42                     0.291637                                11.274967   

    ARG_industry_value_added_%_growth  AUS_gdp_per_capita_%_growth  \
40                          -1.386956                     1.009692   
41                           0.512524                     1.470736   
42                           0.964538                     1.563630   

    AUS_consumer_price_index_%  AUS_unemployment_%_of_total_labor_force  ...  \
40                    1.353727                                 6.468546  ...   
41                    1.739290                                 6.476210  ...   
42                    2.032605                                 6.483084  ...   

    GBR_consumer_price_index_%  GBR_unemployment_%_of_total_labor_force  \
40                    1.457070                                 4.580937   
41                    1.775478                                 4.681121   
42                    1.992303                                 4.773256   

    GBR_government_debt_total_%_of_gdp  GBR_government_expense_%_gdp  \
40                          191.566660                     46.260029   
41                          192.347959                     45.422538   
42                          193.116626                     44.682401   

    GBR_industry_value_added_%_growth  USA_gdp_per_capita_%_growth  \
40                          -2.003782                    -0.276751   
41                          -0.203435                     0.968701   
42                           0.164021                     1.351526   

    USA_consumer_price_index_%  USA_unemployment_%_of_total_labor_force  \
40                    1.925820                                 7.541917   
41                    2.246906                                 7.165148   
42                    2.395839                                 6.885755   

    USA_government_debt_total_%_of_gdp  USA_government_expense_%_gdp  
40                          176.148250                     32.960833  
41                          192.481725                     32.985316  
42                          210.859812                     33.009191  

[3 rows x 91 columns]

In [674]:
# load the model from disk

filename = 'model_1981_upsampled.sav'
loaded_model = pickle.load(open(filename, 'rb'))

filename= 'scaler_1981_upsampled.sav'
loaded_scaler = pickle.load(open(filename, 'rb'))

X_scaled = loaded_scaler.transform(X)

predictions = loaded_model.predict(X_scaled)
print(np.argmax(predictions, axis=1))

1/1 [==============================] - 0s 87ms/step
[2 3 3]


In [ ]:
# 1: Expansion, 2: recession , 3:recovery , 4: downturn

# Conclusion

### The first  model (accuracy 93%) predicted recession for 2021 and 2022 and expansion for 2023.

### This second model (accuracy 81%) predicted recession in 2021 and recovery in 2022 and 2023.